Add Levenshtein distance to ustring class and expose that to GDScript

joao-pedro-braz opened the pull request:
## Description

This PR adds a fast Levenshtein Distance implementation to the String class. The implementation is exposed through the edit_distance method, which has also been made accessible to GDScript.

The Levenshtein distance is a measure of how many edits one String needs to become another. It's similar to the Sorensen-Dice coefficient ('similarity' method in the String class), in the sense that both measure how much one String is similar to another, but Levenshtein is usually more accurate when it comes to error correction (i.e., typos).

The Damerau-Levenshtein (Levenshtein, but with transposition as one of the valid transformations) distance has been implemented for Strings with size less than or equal to 32.

Unit tests have also been added, which compare this implementation against a naive one, resulting in a 100% match.

## Reason

I'm working on a Quick Open refactor, which will make use of this API to suggest better matches for searches. This could be done with the similarity method, but as stated previously, the Levenshtein distance generally works better for the type of suggestions the Quick Open have to do.


## Next step

A good next step would be improving this implementation by computing the Damerau-Levenshtein distance for all lengths.

## Sources

Sources:
  - About the "fast" claim:
    - This implementation is 1.96x faster than the naive approach when comparing 250 randomly generated strings, with a runtime of 5021 vs 9880 msec on a Ryzen 5 5600G.
  - This implementation is based off: https://github.com/ka-weihe/fastest-levenshtein
  - About the error correction claim: https://jacoby.github.io/2019/01/08/levenshtein-sorensendice-and-practical-information-theory.html
  - About the Damerau-Levenshtein Distance: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance
- Damerau-Levenshtein distance implemented according to this paper: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.19.7158&rep=rep1&type=pdf

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

joao-pedro-braz committed:
Add Levenshtein distance to ustring class and expose that to GDScript

This PR adds a fast Levenshtein Distance implementation to the String class. The implementation is exposed through the levenshtein_distance method, which has also been made accessible to GDScript.

The Levenshtein distance is a measure of how many edits one String needs to become another. It's similar to the Sorensen-Dice coefficient ('similarity' method in the String class), in the sense that both measure how much one String is similar to another, but Levenshtein is usually more accurate when it comes to error correction (i.e., typos).

Unit tests have also been added, which compare this implementation against a naive one, resulting in a 100% match.

I'm working on a Quick Open refactor, which will make use of this API to suggest better matches for searches. This could be done with the similarity method, but as stated previously, the Levenshtein distance generally works better for the type of suggestions the Quick Open have to do.

A good next step would be improving this implementation by computing the Damerau-Levenshtein distance, which is similar to the 'original' Levenshtein distance, but with transposition added as a possible operation when computing the Edit Distance.

Sources:
  - About the "fast" claim:
    - This implementation is 1.96x faster than the naive approach when comparing 250 randomly generated strings, with a runtime of 5021 vs 9880 msec on a Ryzen 5 5600G.
  - This implementation is based off: https://github.com/ka-weihe/fastest-levenshtein
  - About the error correction claim: https://jacoby.github.io/2019/01/08/levenshtein-sorensendice-and-practical-information-theory.html
  - About the Damerau-Levenshtein Distance: https://en.wikipedia.org/wiki/Damerau%E2%80%93Levenshtein_distance

Calinou commented:
> but Levenshtein is usually more accurate when it comes to error correction (i.e., typos).

In this case, it makes sense [this part](https://github.com/Calinou/godot/blob/051afb5b61d23f43681a4dbf8b00ab40d3ec7b39/core/input/input_map.cpp#L71) to use Levenshtein distance instead of `similarity()`.

bruvzg commented on core/string/ustring.cpp:
```diff
@@ -3371,6 +3371,145 @@ float String::similarity(const String &p_string) const {
 
+// Based on: https://github.com/ka-weihe/fastest-levenshtein
+int String::levenshtein_distance(const String &p_string) const {
+	auto a = *this;
+	auto b = p_string;
```
We do not use `auto` in the Godot codebase (with a few exceptions), so this should be `const String &`.

bruvzg commented on core/string/ustring.cpp:
```diff
@@ -3371,6 +3371,145 @@ float String::similarity(const String &p_string) const {
+		return a.length();
+	}
+
+	HashMap<char32_t, uint> peq(0x10000);
```
Sizeless `uint` is not a standard C++ type, it should be either `unsigned`/`unsigned int` or `uintXX_t`.

joao-pedro-braz commented:
Did some more work on this:
- Renamed the API from levenshtein_distance to edit_distance
- Implemented Damerau-Levenshtein Distance for Strings with size less than or equal to 32
- Redid the tests by testing against known cases, instead of randomly building test cases

Damerau-Levenshtein Distance implemented according to this paper: https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.19.7158&rep=rep1&type=pdf

The reasoning behind implementing Damerau-Levenshtein Distance only for Strings <= 32 is two fold:
1. bit-vector manipulation on Strings bigger than a single computer-word  (a char32_t) makes the algorithm 20% slower, according to the linked paper, while they're essentially free for smaller Strings.
2. It's not as relevant for bigger Strings anyway.
3. Bonus point, I tried implementing it, but failed :+1:

vnen commented:
I was actually thinking of changing search in all dialogs and code completion to use Levenshtein distance. The Sorensen-Dice is not giving good results.

joao-pedro-braz commented:
> I was actually thinking of changing search in all dialogs and code completion to use Levenshtein distance. The Sorensen-Dice is not giving good results.

I would be happy to work on that, do you have a list with these use cases? Should I just look for usages of String#similarity?

@vnen

joao-pedro-braz commented:
> > but Levenshtein is usually more accurate when it comes to error correction (i.e., typos).
> 
> In this case, it makes sense [this part](https://github.com/godotengine/godot/blob/cd03028915295e67d72cd91dbd46eafb207f92d1/core/input/input_map.cpp#L71) to use Levenshtein distance instead of `similarity()`. This should be done in a separate PR though.

Implemented this on: https://github.com/godotengine/godot/pull/76354

pafuent commented:
@joao-pedro-braz If you don't have time to work on this anymore, I can try to rebase this PR and #76354
If you agree, I'll create a new PRs because I can't update yours. Of course I'll add you as a co author.
